numpy/core/src/umath/ufunc_object.c

```diff
@@ -1354,21 +1389,49 @@ iterator_loop(PyUFuncObject *ufunc,
     for (int i = 0; i < nin; i++) {
         baseptrs[i] = PyArray_BYTES(op_it[i]);
     }
+    if (masked) {
+        baseptrs[nop] = PyArray_BYTES(op_it[nop]);
+    }
     if (NpyIter_ResetBasePointers(iter, baseptrs, NULL) != NPY_SUCCEED) {
         NpyIter_Deallocate(iter);
         return -1;
     }
 
+    /*
+     * Get the inner loop.
+     */
+    int needs_api = 0;
+    PyUFuncGenericFunction innerloop;
+    void *innerloopdata = NULL;
+    if (masked) {
+        if (PyUFunc_DefaultMaskedInnerLoopSelector(ufunc,
+                dtypes, &innerloop, (NpyAuxData **)&innerloopdata,
+                &needs_api) < 0) {
+            NpyIter_Deallocate(iter);
+            return -1;
+        }
+    }
+    else {
+        if (ufunc->legacy_inner_loop_selector(ufunc, dtypes,
+                &innerloop, &innerloopdata, &needs_api) < 0) {
+            NpyIter_Deallocate(iter);
+            return -1;
+        }
+    }
+
     /* Get the variables needed for the loop */
     NpyIter_IterNextFunc *iternext = NpyIter_GetIterNext(iter, NULL);
     if (iternext == NULL) {
         NpyIter_Deallocate(iter);
+        if (masked) {
+            NPY_AUXDATA_FREE((NpyAuxData *)innerloopdata);
+        }
         return -1;
     }
     char **dataptr = NpyIter_GetDataPtrArray(iter);
     npy_intp *strides = NpyIter_GetInnerStrideArray(iter);
     npy_intp *countptr = NpyIter_GetInnerLoopSizePtr(iter);
-    int needs_api = NpyIter_IterationNeedsAPI(iter);
+    needs_api |= NpyIter_IterationNeedsAPI(iter);
```

Comment:
In case anyone wonders: Currently at this point `needs_api` is never set (passing it earlier is a well intentioned no-op).  But the future can see that happening, the concept is good. (Except that `NpyIter_IterationNeedsAPI` is not perfect either right now.)